lib/wopinator/lock.rb

```diff
@@ -0,0 +1,93 @@
+require 'time'
+require 'wopinator/lock_id'
+
+module Wopinator
+  class Lock
+    EXPIRES_IN = 1800.freeze # 30 minutes
+
+    attr_reader :id, :old_id, :timestamp
+
+    def initialize(id, old_id = nil, timestamp = Time.now)
+      self.id = id
+      self.old_id = old_id
+      self.timestamp = timestamp
+    end
+
+    def eql?(lock)
+      return false if invalid? || lock.invalid?
+
+      if old_id.empty? && lock.old_id.valid?
+        id == lock.old_id
+      elsif old_id.valid? && lock.old_id.empty?
+        old_id == lock.id
+      else
+        id == lock.id
+      end
+    end
+
+    def !=(lock)
+      !eql?(lock)
+    end
+
+    def ==(lock)
+      eql?(lock)
+    end
+
+    def set(id)
+      touch!
+      self.id = id
+    end
+
+    def clear!
+      self.timestamp = nil
+      self.id = nil
+      self.old_id = nil
+    end
+
+    def touch!
+      self.timestamp = Time.now
+    end
+
+    def timestamp=(value)
+      @timestamp = parse_timestamp(value)
+    end
+
+    def expired?
+      (timestamp + EXPIRES_IN) < Time.now
+    end
+
+    def empty?
+      id.empty? && old_id.empty?
+    end
+
+    def invalid?
+      !valid?
+    end
+
+    def valid?
+      id.valid? && (old_id.valid? || old_id.empty?) && !expired?
```

Comment:
It looks like we permit `old_id` to be `nil` in the initializer, should we check `old_id.nil?` as part of the grouped expression?